app/src/batch_process.py

```diff
@@ -16,22 +20,34 @@ async def batch_process(file_path: str, engine: ChatEngineInterface) -> str:
         rows = list(reader)  # Convert reader to list to preserve order
         questions = [row["question"] for row in rows]
 
-        # Process questions in parallel while preserving order
+        # Follow example usage https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example
         with ThreadPoolExecutor() as executor:
-            processed_data = list(executor.map(lambda q: _process_question(q, engine), questions))
+            # Process questions in parallel while preserving order
+            futures = [
+                executor.submit(_process_question, i, q, engine)
+                for i, q in enumerate(questions, start=1)
+            ]
+            logger.info("Submitted %i questions for processing", len(futures))
+
+            def update_rows() -> None:
+                logger.info("Waiting for results...")
+                # Update rows with processed data while preserving original order
+                for row, f in zip(rows, futures, strict=True):
+                    row.update(f.result())  # f.result() is a blocking call
+                logger.info("Updated %i rows", len(rows))
 
-        # Update rows with processed data while preserving original order
-        for row, data in zip(rows, processed_data, strict=True):
-            row.update(data)
```

Comment:
The original code can block for longer than 25 seconds, which causes the websocket connection to terminate -- see investigation in [Slack](https://nava.slack.com/archives/C06ETE82UHM/p1736993850761789?thread_ts=1736979278.228369&cid=C06ETE82UHM).